## Reading of CSV files and Manipulation of Data

Link for the dataset: [Click Here](https://www.kaggle.com/pavansubhasht/ibm-hr-analytics-attrition-dataset)

In [ ]:
import pandas as pd

attrition_dataset = pd.read_csv('/content/WA_Fn-UseC_-HR-Employee-Attrition.csv')

In [ ]:
attrition_dataset

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0                    1          2  Life Sciences              1   
1                    8          1  Life Sciences              1   
2                    2          2          Other              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
0                  1  ...                         1            80   
1                  2  ...                         4            80   
2                  4  ...                         2            80   
3                  5  ...                         3            80   
4                  7  ...                         4            80   
...              ...  ...                       ...           ...   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                  1               6                  4   
1                  3              10                  7   
2                  3               0                  0   
3                  3               8                  7   
4                  3               2                  2   
...              ...             ...                ...   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9    

In [ ]:
attrition_dataset.head(10)

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   
5   32        No  Travel_Frequently       1005  Research & Development   
6   59        No      Travel_Rarely       1324  Research & Development   
7   30        No      Travel_Rarely       1358  Research & Development   
8   38        No  Travel_Frequently        216  Research & Development   
9   36        No      Travel_Rarely       1299  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   
5                 2          2  Life Sciences              1               8   
6                 3          3        Medical              1              10   
7                24          1  Life Sciences              1              11   
8                23          3  Life Sciences              1              12   
9                27          3        Medical              1              13   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   
5  ...                         3            80                 0   
6  ...                         1            80                 3   
7  ...                         2            80                 1   
8  ...                         2            80                 0   
9  ...                         2            80                 2   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   
5                  8                      2               2               7   
6                 12                      3               2               1   
7                  1                      2               3               1   
8                 10                      2               3               9   
9                 17                      3               2               7   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  
5                  7                        3                     6  
6                  0                        0                     0  
7                  0                        0                     0  
8                  7                        1             

In [ ]:
attrition_dataset.tail(8)

Age Attrition     BusinessTravel  DailyRate              Department  \
1462   39        No      Travel_Rarely        722                   Sales   
1463   31        No         Non-Travel        325  Research & Development   
1464   26        No      Travel_Rarely       1167                   Sales   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
1462                24          1      Marketing              1   
1463                 5          3        Medical              1   
1464                 5          3          Other              1   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
1462            2056  ...                         1            80   
1463            2057  ...                         2            80   
1464            2060  ...                         4            80   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
1462                 1                 21                      2   
1463                 0                 10                      2   
1464                 0                  5                      2   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
1462               2              20                  9   
1463               3               9                  4   
1464               3               4                  2   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9                  6   
1469               4               4                  3   

      YearsSinceLastPromotion  YearsWithCurrManager  
1462                        9                     6  
1463                        1                     7  
1464                        0                     0  
1465                        0                     3  
1466                        1                     7  
1467                        0                     3  
1468                        0                     8  
1469                        1                     2  

[8 rows x 35 columns]

***read_csv*** is a function in pandas which helps to convert csv format files into dataframe. There are other function present in pandas to read different types of files as well

In [ ]:
# this line of code gives shape of the dataset
attrition_dataset.shape

(1470, 35)

Shape of the dataset gives us a tuple which specifies the number of rows and columns present in the dataset. In the above dataset, there are 1470 rows and 35 columns

In [ ]:
print(attrition_dataset)

      Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0       

head() automatically gives us five rows present at the top of the table. If we want more number of rows we can specify the number within the brackets. 

In [ ]:
initial_10 = attrition_dataset.head(10)
initial_10

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   
5   32        No  Travel_Frequently       1005  Research & Development   
6   59        No      Travel_Rarely       1324  Research & Development   
7   30        No      Travel_Rarely       1358  Research & Development   
8   38        No  Travel_Frequently        216  Research & Development   
9   36        No      Travel_Rarely       1299  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   
5                 2          2  Life Sciences              1               8   
6                 3          3        Medical              1              10   
7                24          1  Life Sciences              1              11   
8                23          3  Life Sciences              1              12   
9                27          3        Medical              1              13   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   
5  ...                         3            80                 0   
6  ...                         1            80                 3   
7  ...                         2            80                 1   
8  ...                         2            80                 0   
9  ...                         2            80                 2   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   
5                  8                      2               2               7   
6                 12                      3               2               1   
7                  1                      2               3               1   
8                 10                      2               3               9   
9                 17                      3               2               7   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  
5                  7                        3                     6  
6                  0                        0                     0  
7                  0                        0                     0  
8                  7                        1             

In [ ]:
initial_10.shape

(10, 35)

In [ ]:
attrition_dataset.head(10)

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   
5   32        No  Travel_Frequently       1005  Research & Development   
6   59        No      Travel_Rarely       1324  Research & Development   
7   30        No      Travel_Rarely       1358  Research & Development   
8   38        No  Travel_Frequently        216  Research & Development   
9   36        No      Travel_Rarely       1299  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   
5                 2          2  Life Sciences              1               8   
6                 3          3        Medical              1              10   
7                24          1  Life Sciences              1              11   
8                23          3  Life Sciences              1              12   
9                27          3        Medical              1              13   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   
5  ...                         3            80                 0   
6  ...                         1            80                 3   
7  ...                         2            80                 1   
8  ...                         2            80                 0   
9  ...                         2            80                 2   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   
5                  8                      2               2               7   
6                 12                      3               2               1   
7                  1                      2               3               1   
8                 10                      2               3               9   
9                 17                      3               2               7   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  
5                  7                        3                     6  
6                  0                        0                     0  
7                  0                        0                     0  
8                  7                        1             

Similarly we have a tail() function as well which gives us the last five rows of the dataframe by default. If you want more values you can specify the number within the brackets

In [ ]:
attrition_dataset.tail(20)

Age Attrition  ... YearsSinceLastPromotion  YearsWithCurrManager
1450   35        No  ...                       1                     7
1451   38        No  ...                       1                     9
1452   50       Yes  ...                       0                     1
1453   36        No  ...                       0                     0
1454   45        No  ...                       0                     1
1455   40        No  ...                       2                     2
1456   35        No  ...                       0                     2
1457   40        No  ...                       0                     2
1458   35        No  ...                       1                     1
1459   29        No  ...                       0                     3
1460   29        No  ...                       0                     4
1461   50       Yes  ...                       2                     0
1462   39        No  ...                       9                     6
1463   31        No  ...                       1                     7
1464   26        No  ...                       0                     0
1465   36        No  ...                       0                     3
1466   39        No  ...                       1                     7
1467   27        No  ...                       0                     3
1468   49        No  ...                       0                     8
1469   34        No  ...                       1                     2

[20 rows x 35 columns]

## 1. Extracting rows and columns

Q1. Extract the attrition column from the dataset and set it to the variable ***response***

In [ ]:
# Method 1
pd.set_option('display.max_rows', None)

response = attrition_dataset.Attrition

response

In [ ]:
# Method 2
response = attrition_dataset['Attrition']
response

0       Yes
1        No
2       Yes
3        No
4        No
       ... 
1465     No
1466     No
1467     No
1468     No
1469     No
Name: Attrition, Length: 1470, dtype: object

Q2. Store the 10th value of Age in age varaible

In [ ]:
# This will give you the 10th value in the dataframe 
age = attrition_dataset['Age'].iloc[9]
age

36

In [ ]:
attrition_dataset.iloc[1466]

Age                                                39
Attrition                                          No
BusinessTravel                          Travel_Rarely
DailyRate                                         613
Department                     Research & Development
DistanceFromHome                                    6
Education                                           1
EducationField                                Medical
EmployeeCount                                       1
EmployeeNumber                                   2062
EnvironmentSatisfaction                             4
Gender                                           Male
HourlyRate                                         42
JobInvolvement                                      2
JobLevel                                            3
JobRole                     Healthcare Representative
JobSatisfaction                                     1
MaritalStatus                                 Married
MonthlyIncome               

Q3. Store first ten values of column Age in **age_list**

In [ ]:
# If you want a range of values you can specify the limit within the iloc brackets. The value before the icon is the
# starting point and the value after the colon is ending point. The ending value is not included. So it will return us 
# elements from index 0 to 9.

age_list = attrition_dataset['Age'].iloc[0:10].values
age_list

array([41, 49, 37, 33, 27, 32, 59, 30, 38, 36])

Q3. Extract first ten rows and two columns, 'Age' and 'Attrition'.

In [ ]:
# iloc[rows_range,column_range]
attrition_dataset.iloc[0:10,0:2].values

array([[41, 'Yes'],
       [49, 'No'],
       [37, 'Yes'],
       [33, 'No'],
       [27, 'No'],
       [32, 'No'],
       [59, 'No'],
       [30, 'No'],
       [38, 'No'],
       [36, 'No']], dtype=object)

Q4. Suppose we want two columns 'Gender' and 'Attrition'. Since they do not lie simultaneously, we will be using loc instead of iloc. iloc takes indexes to give you the results but loc uses labels and hence it is quite useful when you want to extract rows or columns that do not lie together

In [ ]:
columns = ['Gender', 'Attrition']
gender_attrition = attrition_dataset.loc[:10,columns]
gender_attrition

Gender Attrition
0   Female       Yes
1     Male        No
2     Male       Yes
3   Female        No
4     Male        No
5     Male        No
6   Female        No
7     Male        No
8     Male        No
9     Male        No
10    Male        No

Now if i want data at indexes 1, 148 and 1000, I will do the following

In [ ]:
indices = [1, 148, 1000]
gender_attrition_specefic = gender_attrition.loc[indices,:]
gender_attrition_specefic

Gender Attrition
1       Male        No
148     Male        No
1000  Female        No

In [ ]:
# You can also combine the two processes above.

indices = [1, 148, 1000]
columns = ['Gender', 'Attrition']
gender_attrition = gender_attrition.loc[indices,columns]
gender_attrition

Gender Attrition
1       Male        No
148     Male        No
1000  Female        No

## Extracting rows based on conditions

Q5. Extract rows such that the dataframe **males_only** consists only rows with Gender value as Male

In [ ]:
males_only = attrition_dataset[attrition_dataset['Gender'] == 'Male']
males_only.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   
5   32        No  Travel_Frequently       1005  Research & Development   
7   30        No      Travel_Rarely       1358  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
4                 2          1        Medical              1               7   
5                 2          2  Life Sciences              1               8   
7                24          1  Life Sciences              1              11   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
1  ...                         4            80                 1   
2  ...                         2            80                 0   
4  ...                         4            80                 1   
5  ...                         3            80                 0   
7  ...                         2            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
1                 10                      3               3              10   
2                  7                      3               3               0   
4                  6                      3               3               2   
5                  8                      2               2               7   
7                  1                      2               3               1   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
1                  7                        1                     7  
2                  0                        0                     0  
4                  2                        2                     2  
5                  7                        3                     6  
7                  0                        0                     0  

[5 rows x 35 columns]

Q6. Extract rows such that the new dataframe has males that are above 40

In [ ]:
var1 == 1 and var2 == 3

In [ ]:
males_research = attrition_dataset[(attrition_dataset['Gender'] == 'Male') & (attrition_dataset['Age'] > 40)]
males_research.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
1    49        No  Travel_Frequently        279  Research & Development   
27   42        No      Travel_Rarely        691                   Sales   
31   44        No      Travel_Rarely       1459  Research & Development   
36   50       Yes      Travel_Rarely        869                   Sales   
48   46        No  Travel_Frequently       1211                   Sales   

    DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
1                  8          1  Life Sciences              1               2   
27                 8          4      Marketing              1              35   
31                10          4          Other              1              40   
36                 3          2      Marketing              1              47   
48                 5          4      Marketing              1              62   

    ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
1   ...                         4            80                 1   
27  ...                         4            80                 1   
31  ...                         4            80                 0   
36  ...                         3            80                 0   
48  ...                         3            80                 0   

    TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
1                  10                      3               3              10   
27                 10                      2               3               9   
31                  9                      5               4               4   
36                  3                      2               3               3   
48                 14                      4               3               9   

   YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
1                   7                        1                     7  
27                  7                        4                     2  
31                  2                        1                     3  
36                  2                        0                     2  
48                  6                        0                     8  

[5 rows x 35 columns]

In [ ]:
females_only = attrition_dataset[(attrition_dataset['Gender']=='Female') & (attrition_dataset['Age'] > 40)]
females_only.head()

Age Attrition  ... YearsSinceLastPromotion  YearsWithCurrManager
0    41       Yes  ...                       0                     5
6    59        No  ...                       0                     0
18   53        No  ...                       3                     7
25   53        No  ...                       4                     8
28   44        No  ...                       5                    17

[5 rows x 35 columns]

Q7. Find the mean and the median of the Monthly Income

In [ ]:
mean_income = attrition_dataset['MonthlyIncome'].mean()
median_income = attrition_dataset['MonthlyIncome'].median()

mean_income

6502.931292517007

In [ ]:
median_income

4919.0

In [ ]:
attrition_dataset.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000  1470.000000         1470.0   
mean     36.923810   802.485714          9.192517     2.912925            1.0   
std       9.135373   403.509100          8.106864     1.024165            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   802.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  ...  RelationshipSatisfaction  StandardHours  \
count  1470.000000  ...               1470.000000         1470.0   
mean      2.063946  ...                  2.712245           80.0   
std       1.106940  ...                  1.081209            0.0   
min       1.000000  ...                  1.000000           80.0   
25%       1.000000  ...                  2.000000           80.0   
50%       2.000000  ...                  3.000000           80.0   
75%       3.000000  ...                  4.000000           80.0   
max       5.000000  ...                  4.000000           80.0   

       StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
count       1470.000000        1470.000000            1470.000000   
mean           0.793878          11.279592               2.799320   
std            0.852077           7.780782               1.289271   
min            0.000000           0.000000               0.000000   
25%            0.000000           6.000000               2.000000   
50%            1.000000          10.000000               3.000000   
75%            1.000000          15.000000               3.000000   
max            3.000000          40.000000               6.000000   

       WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
count      1470.000000     1470.000000         1470.000000   
mean          2.761224        7.008163            4.229252   
std           0.706476        6.126525            3.623137   
min           1.000000        0.000000            0.000000   
25%           2.000000        3.000000            2.000000   
50%           3.000000        5.000000            3.000000   
75%           3.000000        9.000000            7.000000   
max           4.000000       40.000000           18.000000   

       YearsSinceLastPromotion  YearsWithCurrManager  
count              1470.000000           1470.000000  
mean                  2.187755              4.123129  
std                   3.222430              3.568136  
min                   0.000000              0.000000  
25%                   0.000000              2.000000  
50%                   1.000000              3.000000  
75%                   3.000000              7.000000  
max                  15.000000             17.000000  

[8 rows x 26 columns]

Q8. Check how many employees belong to a particular department 

In [ ]:
attrition_dataset['Department'].value_counts()

Research & Development    961
Sales                     446
Human Resources            63
Name: Department, dtype: int64

In [ ]:
attrition_dataset['Gender'].value_counts()

Male      882
Female    588
Name: Gender, dtype: int64

## Checking missing values

In [ ]:
attrition_dataset.isnull()

Attrition  BusinessTravel  ...  YearsSinceLastPromotion  YearsWithCurrManager
Age                             ...                                               
41       False           False  ...                    False                 False
49       False           False  ...                    False                 False
37       False           False  ...                    False                 False
33       False           False  ...                    False                 False
27       False           False  ...                    False                 False
..         ...             ...  ...                      ...                   ...
36       False           False  ...                    False                 False
39       False           False  ...                    False                 False
27       False           False  ...                    False                 False
49       False           False  ...                    False                 False
34       False           False  ...                    False                 False

[1470 rows x 34 columns]

In [ ]:
# Checks null values
attrition_dataset.isnull().sum()

Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithC

## Renaming columns

Q9. Rename the column 'Attrition' to 'attrition'

In [ ]:
df_2 = attrition_dataset.rename(columns={'Attrition':'attrition', 'Gender': 'male-female'})

In [ ]:
attrition_dataset.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [ ]:
df_2.head()

Age attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [ ]:
attrition_dataset.rename(columns={'Attrition':'attrition'}, inplace=True)  #{Original_col: Rename_col}
attrition_dataset.head()

Age attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

**inplace** by default has False values, i.e, if you don't mention the keyword it will not permanently rename the columns requested. This helps us to retain the original dataframe and we can assign to another variable if we want a dataframe with renamed columns.

You can use inplace=True if you are sure that you will not need those columns later since it permanently renames the requested columns requested. 

## Drop columns from dataframe

Q10. Drop columns **EmployeeCount** and **Over18** from the dataset

In [ ]:
abc = attrition_dataset.drop(columns=['EmployeeCount', 'Over18'], axis=1, inplace=False)

In [ ]:
abc

Q11. Drop rows 1, 48, 100 from the dataset

In [ ]:
attrition_dataset.drop(index=[1,48, 100], axis=0)

Age attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
5      32        No  Travel_Frequently       1005  Research & Development   
6      59        No      Travel_Rarely       1324  Research & Development   
7      30        No      Travel_Rarely       1358  Research & Development   
8      38        No  Travel_Frequently        216  Research & Development   
9      36        No      Travel_Rarely       1299  Research & Development   
10     35        No      Travel_Rarely        809  Research & Development   
11     29        No      Travel_Rarely        153  Research & Development   
12     31        No      Travel_Rarely        670  Research & Development   
13     34        No      Travel_Rarely       1346  Research & Development   
14     28       Yes      Travel_Rarely        103  Research & Development   
15     29        No      Travel_Rarely       1389  Research & Development   
16     32        No      Travel_Rarely        334  Research & Development   
17     22        No         Non-Travel       1123  Research & Development   
18     53        No      Travel_Rarely       1219                   Sales   
19     38        No      Travel_Rarely        371  Research & Development   
20     24        No         Non-Travel        673  Research & Development   
21     36       Yes      Travel_Rarely       1218                   Sales   
22     34        No      Travel_Rarely        419  Research & Development   
23     21        No      Travel_Rarely        391  Research & Development   
24     34       Yes      Travel_Rarely        699  Research & Development   
25     53        No      Travel_Rarely       1282  Research & Development   
26     32       Yes  Travel_Frequently       1125  Research & Development   
27     42        No      Travel_Rarely        691                   Sales   
28     44        No      Travel_Rarely        477  Research & Development   
29     46        No      Travel_Rarely        705                   Sales   
30     33        No      Travel_Rarely        924  Research & Development   
31     44        No      Travel_Rarely       1459  Research & Development   
32     30        No      Travel_Rarely        125  Research & Development   
33     39       Yes      Travel_Rarely        895                   Sales   
34     24       Yes      Travel_Rarely        813  Research & Development   
35     43        No      Travel_Rarely       1273  Research & Development   
36     50       Yes      Travel_Rarely        869                   Sales   
37     35        No      Travel_Rarely        890                   Sales   
38     36        No      Travel_Rarely        852  Research & Development   
39     33        No  Travel_Frequently       1141                   Sales   
40     35        No      Travel_Rarely        464  Research & Development   
41     27        No      Travel_Rarely       1240  Research & Development   
42     26       Yes      Travel_Rarely       1357  Research & Development   
43     27        No  Travel_Frequently        994                   Sales   
44     30        No  Travel_Frequently        721  Research & Development   
45     41       Yes      Travel_Rarely       1360  Research & Development   
46     34        No         Non-Travel       1065                   Sales   
47     37        No      Travel_Rarely        408  Research & Development   
49     35        No      Travel_Rarely       1229  Research & Development   
50     48       Yes      Travel_Rarely        626  Research & Development   
51     28       Yes      Travel_Rarely       1434  Research & Development   
52     44        No      Travel_Rarely       1488                   Sales   
53

**inplace** by default has False values, i.e, if you don't mention the keyword it will not permanently delete the columns/rows requested. This helps us to retain the original dataframe and we can assign to another variable if we want a dataframe without those columns/rows.

You can use inplace=True if you are sure that you will not need those columns later since it permanently deletes the requested columns or rows requested. 

You can also retrieve the unique columns in the dataframe

In [ ]:
attrition_dataset['Attrition'].unique()

array(['Yes', 'No'], dtype=object)

In [ ]:
attrition_dataset.columns.unique()

Index(['Age', 'attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

You can also retrieve unique values present in the columns

In [ ]:
attrition_dataset['Department'].unique()

array(['Sales', 'Research & Development', 'Human Resources'], dtype=object)

In [ ]:
attrition_dataset.to_csv('test.csv')